### PR TITLE
chore(nix): make it faster to get melange

### PR DESCRIPTION
**Pull request by anmonteiro**
- [nix-overlays](https://github.com/nix-ocaml/nix-overlays) enables flambda by default
- depending on the melange flake forces dune developers to build an extra flambda compiler to work on Dune
- this change uses the upstream OCaml from nixpkgs which should always be cached in cache.nixos.org

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -33,7 +34,31 @@
           cinaps = "*";
           ocamlfind = "1.9.2";
         };
-        pkgs = nixpkgs.legacyPackages.${system};
+        pkgs = nixpkgs.legacyPackages.${system}.extend (self: super: {
+          ocaml-ng = super.ocaml-ng // {
+            ocamlPackages_4_14 =
+              super.ocaml-ng.ocamlPackages_4_14.overrideScope' (oself: osuper: {
+                dune_3 = osuper.dune_3.overrideAttrs (o: {
```
Comment (anmonteiro) [inline flake.nix:41]: should only be needed until https://github.com/NixOS/nixpkgs/pull/196899 is merged.

**Review by Alizter — APPROVED** [review]
Thanks!

**Comment by rgrinberg** [comment]
@anmonteiro instead of waiting for upstream, could you add an overlay here?

File: flake.nix [hunk 2]
```diff
@@ -12,6 +12,7 @@
       flake = false;
     };
     melange.url = "github:melange-re/melange";
+    melange.inputs.nixpkgs.follows = "nixpkgs";
```
Comment (rgrinberg) [inline flake.nix:15]: instead of reusing our own nixpkgs, would it be possible to just insert an overlay into melange's nixpkgs?

**Comment by anmonteiro** [comment]
> @anmonteiro instead of waiting for upstream, could you add an overlay here?

I like that suggestion. will do

File: flake.nix (hunk 2 above)
Comment (anmonteiro) [reply flake.nix:15]: If I understood you correctly, I don't think that's a good idea. Melange's nixpkgs are my nix overlays, which is what's causing the issue of having 2 OCaml compilers in the first place (clambda and flambda).

Comment (rgrinberg) [reply flake.nix:15]: Right but we can override your overlay? We don't need a fast melange since we only run it in trivial test projects.

Comment (rgrinberg) [reply flake.nix:15]: I guess if our goal is to speed up the build of melange this is ideal.

I'm just worried that if you decide to modify or overlay your nixpkgs somehow, melange won't build in our dev shell.

Anyway it's enough to keep an eye out for this for now.

Comment (anmonteiro) [reply flake.nix]: Alright, I removed the `follows` line. It doesn't impact the build: we just download an extra tarball of `nixpkgs`.

**Comment by anmonteiro** [comment]
Switched this to use a melage-provided overlay (added in https://github.com/melange-re/melange/pull/430).

File: flake.nix [hunk 3]
```diff
@@ -33,7 +33,27 @@
...
+        pkgs = nixpkgs.legacyPackages.${system}.appendOverlays [
+          (self: super: {
+            ocaml-ng = super.ocaml-ng // {
+              ocamlPackages_4_14 =
+                super.ocaml-ng.ocamlPackages_4_14.overrideScope' (oself: osuper: {
+                  melange-compiler-libs = with oself; buildDunePackage rec {
+                    pname = "melange-compiler-libs";
+                    version = "0.0.1-414";
+                    src = builtins.fetchurl {
+                      url = "https://github.com/melange-re/melange-compiler-libs/releases/download/${version}/melange-compiler-libs-${version}.tbz";
```
Comment (rgrinberg) [inline flake.nix:45]: Would it be possible to add melange-compiler-libs as a flake input instead?

Comment (anmonteiro) [reply flake.nix]: done in https://github.com/ocaml/dune/pull/6347/commits/76eab97a04049f68d63a902d4c7ecbdf590a7bd7

**Comment by anmonteiro** [comment]
Diff looks a lot better with https://github.com/ocaml/dune/pull/6347/files?w=1

**Pull request merged**